tests/codegen/slice-iter-provenance.rs

```diff
@@ -0,0 +1,20 @@
+//@ compile-flags: -O
+#![crate_type = "lib"]
+
+// Back when `slice::Iter(Mut)` carried two pointers, LLVM was not *allowed*
+// to optimize out certain things at the IR level since starting from the other
+// end had a difference provenance and thus wasn't actually equivalent.
+
+// Now that they're `{ ptr, usize }`, however, there's only one provenance used
+// by everything producing pointers
+
+// FIXME: Add more tests here LLVM has fixed the following bug:
+// <https://github.com/llvm/llvm-project/issues/86417>
+
+// CHECK-LABEL: @first_via_nth_back
+#[no_mangle]
+pub unsafe fn first_via_nth_back(mut it: std::slice::Iter<'_, i8>) -> &i8 {
```

Comment:
how realistic is this test though? often the iter construction gets inlined too which means llvm sees they come from the same pointer

https://rust.godbolt.org/z/Pj44sTbYf

Reply:
Hrm, yes, I can see how this is bad for single-access functions like that. Otoh I'm concerned about how things optimize in loops where even if the provenance of the pointers is potentially-different it only gets accessed once (to calculate the length) and then it operates off only one pointer.